vembrane/modules/filter.py

```diff
@@ -160,67 +166,127 @@ def filter_vcf(
+                    raise ValueError(
+                        f"Filtering of BND records with multiple mates is unsupported "
+                        f"(see VCF 4.3, section 5.4.3 'Multiple mates'):\n{str(record)}"
+                    )
+
+                # if we have a mate pair, use their sorted ids as a key
+                mate_id = mate_ids[0] if len(mate_ids) == 1 else None
+                mate_pair = mate_key([record.id, mate_id]) if mate_ids else None
+
+                # if EVENT is set, it has priority over MATEID.
+                event_name = event_name or mate_pair
+
+                # if both EVENT and MATEID are not set
+                # we can only treat this BND record as a regular one,
+                # (or somehow try to guess its correct mate)
+                if not event_name:
+                    print(
+                        f"Warning: Encountered breakend record at index {idx} "
+                        f"without either of MATEID or EVENT specified, "
+                        f"treating it as a regular record:\n{str(record)}",
+                        file=stderr,
+                    )
+                    if record_has_passed:
                         yield record
+                    continue
+
+                event = events.get(event_name, None)
+
+                # if there's already an associated event
+                if event:
+                    # add this record to the event
+                    event.add(record, record_has_passed)
+
+                    # if we already know that the event is a "PASS"…
+                    if event.passed:
+                        # … emit all records associated with it
+                        yield from event.emit()
+
+                        # in the case of a simple mate pair, we can delete the event
+                        # at this point, because no more records will be added to it
+                        if event.is_mate_pair():
+                            del events[mate_pair]
+                else:
+                    # if there's no entry for the event or mate pair yet, create one
+                    is_mate_pair = mate_pair and mate_pair == event_name
+                    event = BreakendEvent(event_name, is_mate_pair)
+                    event.add(record, record_has_passed)
+                    events[event_name] = event
+            elif record_has_passed:
+                yield record
+
+        if len(events) > 0:
+            # output BNDs if any are were left unprocessed
+            for event_name, event in events.items():
+                if event_name and event.passed:
+                    yield from event.emit()
     else:
+
+        # There may be BND record which have neither of
+        # INFO['EVENT'], INFO['MATEID'] or even ID specified.
+        # In that case, we still need a pseudo-event name,
+        # which we construct from the record's index in the file.
+        def get_event_name(record: VariantRecord) -> str:
```

Comment:
Why isn't this function also used in the above `if not preserve_order` branch? I'm probably missing a subtle difference here, but just want to make sure this is intended.

Reply:
There is indeed a minor difference as to how exactly records are filtered / "pass" information is memorized here:

In the case where order does not matter, we keep the records in memory until we have sufficient information.
Records are always associated to a pseudo-event, which is either defined through the EVENT tag or two MATE_IDs combined.
However, in the case neither of those are set, the information in the VCF regarding breakends is basically under-specified (it's not trivial to link the correct breakend records to form mates etc without the mate id information etc), so we treat them as regular non-BND records (which corresponds to the `None` case in `event or mate_pair or None`).

When order does matter, the way we keep track of which events to keep/discard changes slightly:
we just keep a set of event names which we populate in a first pass (i.e. if a record passes the expression and belongs to some event, that event is added to the set).
In a second pass, we check if the event name was memorized during the first pass, and if so, output that record.
In the case where both EVENT and mate-pair information are unavailable, we still need store *some* key/event/name in the `events` set, and that key is then either the record's ID or simply the index in the file.

Reply:
Thanks for the detailed info. I'm just worried, that code might diverge in future edits and lead to inconsistent behaviour. So maybe it makes sense to reduce this to a function that just returns `(event_name, mate_id)` and then have the different handling on those results?

Reply:
That's a good suggestion, thanks, implemented now ;)